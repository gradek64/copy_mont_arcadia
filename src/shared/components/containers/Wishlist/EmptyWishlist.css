/** @define EmptyWishlist */
.EmptyWishlist {
  padding: 20px 10px;

  &-wishlistIcon {
    background-image: $wishlist-header-icon;
    background-repeat: no-repeat;
    background-size: cover;
    height: $wishlist-empty-state-icon-height;
    width: $wishlist-empty-state-icon-width;
    display: inline-block;
    position: relative;
    top: 4px;
  }

  &-continueButton {
    margin: 15px auto;
    max-width: 250px;
  }

  &-signInButton {
    text-transform: uppercase;
    text-decoration: underline;
  }
}
