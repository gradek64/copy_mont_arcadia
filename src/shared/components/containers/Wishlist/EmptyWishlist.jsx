import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { browserHistory } from 'react-router'

import Button from '../../common/Button/Button'

const WISHLIST_ICON = `<i class='EmptyWishlist-wishlistIcon'></i>`

class EmptyWishlist extends Component {
  static propTypes = {
    isUserAuthenticated: PropTypes.bool,
    onSignInHandler: PropTypes.func,
    visited: PropTypes.array.isRequired,
  }

  static contextTypes = {
    l: PropTypes.func,
  }

  continueShoppingHandler = () => {
    const { visited } = this.props
    const excludedRoutes = [
      '/wishlist',
      '/login',
      '/my-account',
      '/order-complete',
    ]
    if (visited.length <= 1) {
      browserHistory.replace('/')
      return
    }
    const prevRoute = visited[visited.length - 2] // gets prev routes for array lengths 2 and over
    const checkExcluded = excludedRoutes.filter((route) =>
      prevRoute.includes(route)
    )
    return checkExcluded.length > 0
      ? browserHistory.replace('/')
      : browserHistory.goBack()
  }

  render() {
    const { l } = this.context
    const { isUserAuthenticated, onSignInHandler } = this.props

    return (
      <div className="EmptyWishlist">
        <span
          className="EmptyWishlist-message"
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{
            __html: l`Select the ${WISHLIST_ICON} icon next to our products to save them to your Wishlist.`,
          }}
        />
        <Button
          className="EmptyWishlist-continueButton"
          clickHandler={this.continueShoppingHandler}
        >
          {l`Continue Shopping`}
        </Button>
        {!isUserAuthenticated && (
          <span className="EmptyWishlist-message">
            {l`Or `}
            <button
              className="EmptyWishlist-signInButton"
              onClick={onSignInHandler}
            >
              {l`Sign In`}
            </button>
            {l` to retrieve previous Wishlists.`}
          </span>
        )}
      </div>
    )
  }
}

export default EmptyWishlist
