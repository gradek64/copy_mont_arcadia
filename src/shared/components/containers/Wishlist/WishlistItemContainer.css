/** @define WishlistItemContainer */

.WishlistItemContainer {
  position: relative;
  margin: 30px 0 0;
  padding: 0 5px;
  display: flex;
  flex-flow: row wrap;
  width: 100%;
}

/* postcss-bem-linter: ignore */
.WishlistItemContainer .fade-exit-active {
  opacity: 1;
}

/* postcss-bem-linter: ignore */
.WishlistItemContainer .fade-exit {
  opacity: 0.01;
  transition: opacity 400ms ease-in;
}
