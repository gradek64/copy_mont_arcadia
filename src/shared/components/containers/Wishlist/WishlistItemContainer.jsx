import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { TransitionGroup, CSSTransition } from 'react-transition-group'

/**
 * Components
 */
import WishlistItem from './WishlistItem'
import ProductQuickview from '../ProductQuickview/ProductQuickview'
import AddToBagModal from '../../common/AddToBagModal/AddToBagModal'

/**
 * Actions
 */
import { showModal } from '../../../actions/common/modalActions'
import {
  setProductQuickview,
  setProductIdQuickview,
} from '../../../actions/common/quickviewActions'
import { addToBagWithCatEntryId } from '../../../actions/common/shoppingBagActions'
import {
  removeProductFromWishlist,
  captureWishlistEvent,
} from '../../../actions/common/wishlistActions'
import { preserveScroll } from '../../../actions/common/infinityScrollActions'

/**
 * Selectors
 */
import {
  getBaseUrlPath,
  getBrandName,
} from '../../../selectors/configSelectors'

/**
 * Helpers
 */
import { isIE11 } from '../../../lib/browser'

@connect(
  (state) => ({
    baseUrlPath: getBaseUrlPath(state),
    brandName: getBrandName(state),
  }),
  {
    preserveScroll,
    setProductQuickview,
    setProductIdQuickview,
    showModal,
    addToBagWithCatEntryId,
    captureWishlistEvent,
    removeProductFromWishlist,
  }
)
class WishlistItemContainer extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      selectedCatEntryIds: { ...this.getOneSizeOnlyProducts(0, props.items) },
      sizeValidationErrorIds: [],
      isAddingToBag: { item: null, loading: false },
    }
  }

  static propTypes = {
    brandName: PropTypes.string.isRequired,
    grid: PropTypes.number.isRequired,
    items: PropTypes.array.isRequired,
    baseUrlPath: PropTypes.string,
    preserveScroll: PropTypes.func.isRequired,
    setProductQuickview: PropTypes.func.isRequired,
    setProductIdQuickview: PropTypes.func.isRequired,
    showModal: PropTypes.func.isRequired,
    addToBagWithCatEntryId: PropTypes.func.isRequired,
    captureWishlistEvent: PropTypes.func.isRequired,
    removeProductFromWishlist: PropTypes.func.isRequired,
  }

  getOneSizeOnlyProducts = (lastIndex, products = []) => {
    const currentProducts = products.slice(lastIndex)
    return currentProducts.reduce((acc, product) => {
      if (
        Array.isArray(product.sizeAndQuantity) &&
        product.sizeAndQuantity.length === 1
      )
        return { ...acc, [product.parentProductId]: product.catEntryId }

      return acc
    }, [])
  }

  componentWillReceiveProps(nextProps) {
    const { items: currentItems } = this.props
    const { items: newItems } = nextProps
    if (newItems.length !== currentItems.length) {
      this.setState({
        selectedCatEntryIds: {
          ...this.state.selectedCatEntryIds,
          ...this.getOneSizeOnlyProducts(currentItems.length, newItems),
        },
      })
    }
  }

  handleQuickView = (productId) => {
    const { setProductQuickview, setProductIdQuickview, showModal } = this.props
    // We explicitly set the product quick view state to an empty object here
    // in order to aid our analytics decorator on ProductQuickView to know
    // that a product is being quick-viewed. This helps with cases where you
    // quick view the same product sequentially.
    setProductQuickview({})
    setProductIdQuickview(productId)

    const html = <ProductQuickview fromWishlist />
    showModal(html, { mode: 'plpQuickview' })
  }

  handleAddToBag = (e, productId, productDetails) => {
    e.preventDefault()
    const { selectedCatEntryIds } = this.state
    const {
      addToBagWithCatEntryId,
      captureWishlistEvent,
      removeProductFromWishlist,
    } = this.props
    const catEntryId = selectedCatEntryIds[productId]

    if (!selectedCatEntryIds[productId]) {
      this.setState({
        sizeValidationErrorIds: [
          ...this.state.sizeValidationErrorIds,
          productId,
        ],
      })
      return false
    }

    this.setState({ isAddingToBag: { item: productId, loading: true } })

    addToBagWithCatEntryId(catEntryId, <AddToBagModal />)
      .then(() =>
        removeProductFromWishlist({
          productId,
          modifier: 'wishlist',
          productDetails,
          reportToGA: false,
        })
      )
      .then(() =>
        captureWishlistEvent('GA_ADD_TO_BAG_FROM_WISHLIST', {
          productId,
          ...productDetails,
        })
      )
      .finally(() => {
        this.setState({ isAddingToBag: { item: null, loading: false } })
      })
  }

  handleSizeChange = (e, productId) => {
    const catEntryId = e.target.value
    const newValidatedIds = this.state.sizeValidationErrorIds.filter(
      (e) => e !== productId
    )
    this.setState({
      selectedCatEntryIds: {
        ...this.state.selectedCatEntryIds,
        [productId]: catEntryId,
      },
      sizeValidationErrorIds: newValidatedIds,
    })
  }

  handleRemoveFromWishlist = (productId, productDetails) => {
    this.props.removeProductFromWishlist({
      productId,
      modifier: 'wishlist',
      productDetails,
      reportToGA: true,
    })
  }

  handleLinkClick = () => {
    const scrollYPos = window.scrollY || window.pageYOffset
    if (isIE11()) this.props.preserveScroll(scrollYPos)
  }

  renderItems = (grid, items, baseUrlPath, brandName) => {
    return (
      items &&
      items.map((item) => {
        const validateItem = this.state.sizeValidationErrorIds.includes(
          item.parentProductId
        )

        return (
          <CSSTransition key={item.catEntryId} timeout={400} classNames="fade">
            <WishlistItem
              brandName={brandName}
              item={item}
              grid={grid}
              handleAddToBag={this.handleAddToBag}
              handleQuickView={this.handleQuickView}
              handleSizeChange={this.handleSizeChange}
              selectedCatEntryId={
                this.state.selectedCatEntryIds[item.parentProductId]
              }
              baseUrlPath={baseUrlPath}
              handleRemoveFromWishlist={this.handleRemoveFromWishlist}
              sizeValidationError={validateItem}
              handleLinkClick={this.handleLinkClick}
              isAddingToBag={this.state.isAddingToBag}
            />
          </CSSTransition>
        )
      })
    )
  }

  render() {
    const { grid, items, baseUrlPath, brandName } = this.props
    return (
      <ul className="WishlistItemContainer">
        <TransitionGroup className="WishlistItemContainer WishlistItemContainer-animation">
          {this.renderItems(grid, items, baseUrlPath, brandName)}
        </TransitionGroup>
      </ul>
    )
  }
}

export default WishlistItemContainer
