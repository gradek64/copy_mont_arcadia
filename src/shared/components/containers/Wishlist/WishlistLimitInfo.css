/** @define WishlistLimitInfo */

.WishlistLimitInfo {
  &-container {
    width: 100%;
    max-width: 404px;
    padding: 9px 14px 14px 14px;
    border: 1px solid #000;
    margin: 0 auto 40px;
    display: block;
  }

  &-outerWrapper {
    margin: 10px 15px;

    &--withMarginTop {
      margin-top: 40px;
    }
  }
}
