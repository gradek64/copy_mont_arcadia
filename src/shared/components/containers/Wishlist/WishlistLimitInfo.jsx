import React from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

if (process.browser) require('./WishlistLimitInfo.css')

const WishlistLimitInfo = ({ withMarginTop }, { l }) => {
  const outerClassName = cn('WishlistLimitInfo-outerWrapper', {
    'WishlistLimitInfo-outerWrapper--withMarginTop': withMarginTop,
  })
  return (
    <div className={outerClassName}>
      <div className="WishlistLimitInfo-container">
        <p className="WishlistLimitInfo-text">
          {l`You have reached the limit of products allowed in your Wishlist. To add more products, please delete items from your list.`}
        </p>
      </div>
    </div>
  )
}

WishlistLimitInfo.propTypes = {
  withMarginTop: PropTypes.bool,
}

WishlistLimitInfo.defaultProps = {
  withMarginTop: false,
}

WishlistLimitInfo.contextTypes = {
  l: PropTypes.func,
}

export default WishlistLimitInfo
