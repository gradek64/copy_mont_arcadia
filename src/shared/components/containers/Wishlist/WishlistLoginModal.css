/** @define WishlistLoginModal */

.WishlistLoginModal {
  &-forgetPassword {
    margin-bottom: 10px;
  }

  &-forms {
    display: flex;
    flex-direction: column;
  }

  &-header {
    margin-bottom: 20px;
  }

  &-heartIcon {
    display: block;
    background-image: $wishlist-header-icon;
    background-repeat: no-repeat;
    background-size: contain;
    background-position: center;
    margin: 10px auto;
    width: 40px;
    height: 36px;
  }

  &-loginForm,
  &-register {
    margin-bottom: 25px;
    margin-top: 0;
  }

  &-text {
    font-size: 18px;
    line-height: 1.22;
    text-align: center;
  }
}

@media $min-tablet {
  .WishlistLoginModal {
    &-formContainer {
      width: 50%;
      overflow:hidden;
    }

    &-forms {
      flex-direction: row;
      justify-content: center;
    }

    &-header {
      align-items: center;
      display: flex;
      flex-direction: row;
      justify-content: center;
    }

    &-heartIcon {
      margin: 0 20px;
      width: 23px;
      height: 21px;
    }
  }
}

@media $min-laptop {
  .WishlistLoginModal {
    &-formContainer {
      padding: 0 50px;
    }
  }
}
