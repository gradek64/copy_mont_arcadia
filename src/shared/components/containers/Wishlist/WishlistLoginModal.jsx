import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import Login from '../Login/Login'
import Register from '../Register/Register'
import ForgetPassword from '../ForgetPassword/ForgetPassword'

import {
  addToWishlistAfterLogin,
  captureWishlistEvent,
  getPaginatedWishlist,
} from '../../../actions/common/wishlistActions'
import {
  closeModal,
  clearModalChildren,
} from '../../../actions/common/modalActions'

import { getRoutePathWithParams } from '../../../selectors/routingSelectors'

if (process.browser) {
  require('./WishlistLoginModal.css')
}

@connect(
  (state) => ({
    relativePath: getRoutePathWithParams(state),
    isModalCancelled: state.modal.cancelled,
    pendingProductId: state.wishlist.pendingProductId,
  }),
  {
    addToWishlistAfterLogin,
    closeModal,
    getPaginatedWishlist,
    clearModalChildren,
    captureWishlistEvent,
  }
)
class WishlistLoginModal extends Component {
  static propTypes = {
    modifier: PropTypes.string,
    relativePath: PropTypes.string,
    pendingProductId: PropTypes.number,
    addToWishlistAfterLogin: PropTypes.func.isRequired,
    captureWishlistEvent: PropTypes.func,
    afterAddToWishlist: PropTypes.func,
    onCancelLogin: PropTypes.func,
    productDetails: PropTypes.object.isRequired,
  }

  static contextTypes = {
    l: PropTypes.func,
  }

  componentDidMount() {
    const { modifier, captureWishlistEvent } = this.props

    captureWishlistEvent('GA_WISHLIST_MODAL_PAGE_VIEW', { modifier })
  }

  componentWillUnmount() {
    const {
      modifier,
      isModalCancelled,
      captureWishlistEvent,
      onCancelLogin,
    } = this.props
    if (isModalCancelled) {
      captureWishlistEvent('GA_WISHLIST_MODAL_CLOSE', { modifier })
      if (onCancelLogin) onCancelLogin()
    }
  }

  handleSuccess = (authType) => {
    const { modifier, captureWishlistEvent } = this.props

    if (authType) {
      const action = `GA_WISHLIST_${authType.toUpperCase()}_SUCCESS`
      captureWishlistEvent(action)
    }

    switch (modifier) {
      case 'wishlist':
        return this.wishlistPageHandler()
      case 'plp':
      case 'pdp':
      case 'quickview':
      case 'minibag':
      case 'bundle':
        return this.addToWishlistHandler()
      default:
        return this.loginUserHandler()
    }
  }

  handleError = (authType) => {
    const { captureWishlistEvent } = this.props
    const action = `GA_WISHLIST_${authType.toUpperCase()}_ERROR`

    captureWishlistEvent(action)
  }

  loginUserHandler = () => {
    const { closeModal, clearModalChildren } = this.props
    const actions = [closeModal(), clearModalChildren()]
    return Promise.all([actions])
  }

  wishlistPageHandler = () => {
    const { closeModal, clearModalChildren, getPaginatedWishlist } = this.props
    const actions = [closeModal(), clearModalChildren(), getPaginatedWishlist()]
    return Promise.all([actions])
  }

  addToWishlistHandler = () => {
    const {
      addToWishlistAfterLogin,
      afterAddToWishlist,
      clearModalChildren,
      closeModal,
      modifier,
      pendingProductId,
      productDetails,
    } = this.props
    const actions = [clearModalChildren(), closeModal()]

    return addToWishlistAfterLogin(
      pendingProductId,
      modifier,
      productDetails
    ).then(() => {
      return Promise.all(actions).then(() => {
        if (afterAddToWishlist) afterAddToWishlist()
      })
    })
  }

  render() {
    const { l } = this.context
    const { relativePath } = this.props

    return (
      <div className="WishlistLoginModal">
        <div className="WishlistLoginModal-header">
          <i className="WishlistLoginModal-heartIcon" />
          <header className="WishlistLoginModal-text">{l`Sign in to save products to your Wishlist`}</header>
        </div>
        <div className="WishlistLoginModal-forms">
          <div className="WishlistLoginModal-formContainer">
            <Login
              className="WishlistLoginModal-loginForm"
              successCallback={() => this.handleSuccess('login')}
              errorCallback={() => this.handleError('login')}
              formName="wishlistLoginModal"
            />
            <ForgetPassword className="WishlistLoginModal-forgetPassword" />
          </div>
          <div className="WishlistLoginModal-formContainer">
            <Register
              className="WishlistLoginModal-register"
              source="WISHLIST"
              successCallback={() => this.handleSuccess('register')}
              errorCallback={() => this.handleError('register')}
              getNextRoute={() => relativePath}
              formName="wishlistLoginModal"
            />
          </div>
        </div>
      </div>
    )
  }
}

export default WishlistLoginModal
