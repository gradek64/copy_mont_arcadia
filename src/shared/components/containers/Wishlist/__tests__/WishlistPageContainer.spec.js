import { identity, compose, range } from 'ramda'
import { CSSTransition } from 'react-transition-group'

import testComponentHelper, {
  analyticsDecoratorHelper,
  buildComponentRender,
  mountRender,
} from '../../../../../../test/unit/helpers/test-component'

import { MAX_WISHLIST_ITEMS } from '../../../../constants/wishlistConstants'

import DecoratedWishlistPageContainer, {
  WishlistPageContainer,
} from '../WishlistPageContainer'
import WishlistLoginModal from '../../../containers/Wishlist/WishlistLoginModal'
import WishlistLimitInfo from '../WishlistLimitInfo'
import EmptyWishlist from '../EmptyWishlist'
import Loader from '../../../common/Loader/Loader'
import WishlistItemContainer from '../WishlistItemContainer'

const createFakeItem = (i) => ({
  catEntryId: `316218${i}`,
})
const createFakeItems = (quantity) =>
  range(1, quantity + 1).map((i) => createFakeItem(i))
const itemsMock = [createFakeItem('90'), createFakeItem('47')]

const checkTransitionWrapperState = (wrapper, element, isShown) => {
  const transitionWrapper = wrapper.find(element).parent()
  expect(transitionWrapper.is(CSSTransition)).toBe(true)
  expect(transitionWrapper.prop('in')).toBe(isShown)
}

const props = {
  totalItems: 0,
  items: null,
  isLoadingItems: false,
  loadedPages: 1,
  isUserAuthenticated: false,
  location: { action: 'PUSH' },
  hasReachedPageBottom: false,
  isWishlistSync: false,
  getPaginatedWishlist: jest.fn(),
  triggerWishlistLoginModal: jest.fn(),
  preservedScroll: 3,
  visited: [],
}

describe(WishlistPageContainer.name, () => {
  const context = {
    l: jest.fn(identity),
  }

  const renderComponent = testComponentHelper(WishlistPageContainer, {
    context,
  })

  describe('@decorators', () => {
    analyticsDecoratorHelper(DecoratedWishlistPageContainer, 'wishlist', {
      componentName: 'WishlistPageContainer',
      isAsync: false,
    })
  })

  describe('@renders', () => {
    describe('on SSR', () => {
      let oldProcessBrowser
      beforeAll(() => {
        oldProcessBrowser = process.browser
        process.browser = false
      })
      afterAll(() => {
        process.browser = oldProcessBrowser
      })

      it('should render the header', () => {
        const { wrapper } = renderComponent(props)
        expect(wrapper.find('h1').text()).toBe('My Wishlist')
      })

      describe('if the user is logged in and has WL items', () => {
        it('should render the totalItems and a loading spinner', () => {
          const { wrapper } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            totalItems: 14,
          })
          expect(wrapper.find('.WishlistPageContainer-itemCount').text()).toBe(
            '14 items'
          )
          expect(wrapper.find(Loader)).toHaveLength(1)
          checkTransitionWrapperState(wrapper, EmptyWishlist, false)
        })
      })

      describe('if the user is logged in but does not have WL items', () => {
        it('should render the empty WL component', () => {
          const { wrapper } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            totalItems: 0,
          })
          checkTransitionWrapperState(wrapper, EmptyWishlist, true)
          const emptyWL = wrapper.find(EmptyWishlist)
          expect(emptyWL.prop('isUserAuthenticated')).toBe(true)

          expect(wrapper.find(Loader)).toHaveLength(0)
          expect(wrapper.find('WishlistPageContainer-itemCount')).toHaveLength(
            0
          )
        })
      })

      describe('if user is not logged in', () => {
        it('should render the empty WL component', () => {
          const { wrapper } = renderComponent({
            ...props,
            isUserAuthenticated: false,
          })
          checkTransitionWrapperState(wrapper, EmptyWishlist, true)
          const emptyWL = wrapper.find(EmptyWishlist)
          expect(emptyWL.prop('isUserAuthenticated')).toBe(false)

          expect(wrapper.find(Loader)).toHaveLength(0)
          expect(wrapper.find('WishlistPageContainer-itemCount')).toHaveLength(
            0
          )
        })
      })
    })

    describe('on client', () => {
      let oldProcessBrowser
      beforeAll(() => {
        oldProcessBrowser = process.browser
        process.browser = true
      })
      afterAll(() => {
        process.browser = oldProcessBrowser
      })

      describe('when user is not authenticated', () => {
        it('renders <EmptyWishlist /> CSSTransition', () => {
          const { wrapper, getTree } = renderComponent({
            ...props,
            isUserAuthenticated: false,
          })

          checkTransitionWrapperState(wrapper, EmptyWishlist, true)
          checkTransitionWrapperState(wrapper, WishlistItemContainer, false)
          expect(wrapper.find('WishlistPageContainer-itemCount')).toHaveLength(
            0
          )

          expect(getTree()).toMatchSnapshot()
        })
      })

      describe('when user is authenticated but has no wishlisted items', () => {
        it('renders <EmptyWishlist /> CSSTransition', () => {
          const { wrapper, getTree } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            items: [],
            totalItems: 0,
          })

          checkTransitionWrapperState(wrapper, EmptyWishlist, true)
          checkTransitionWrapperState(wrapper, WishlistItemContainer, false)
          expect(wrapper.find('WishlistPageContainer-itemCount')).toHaveLength(
            0
          )

          expect(getTree()).toMatchSnapshot()
        })
      })

      describe('when user has items in the wishlist', () => {
        it('should render the item count', () => {
          const { wrapper } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            items: itemsMock,
            totalItems: 88,
          })
          expect(wrapper.find('.WishlistPageContainer-itemCount').text()).toBe(
            '88 items'
          )
        })

        it('renders <WishlistItemContainer /> CSSTransition', () => {
          const { wrapper, getTree } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            items: itemsMock,
            totalItems: 100,
          })

          checkTransitionWrapperState(wrapper, EmptyWishlist, false)
          checkTransitionWrapperState(wrapper, WishlistItemContainer, true)

          expect(getTree()).toMatchSnapshot()
        })

        it('should render a loader when `isLoadingItems` true', () => {
          const { wrapper } = renderComponent({
            ...props,
            isUserAuthenticated: true,
            items: itemsMock,
            totalItems: 20,
            isLoadingItems: true,
          })

          expect(wrapper.find(Loader)).toHaveLength(1)
        })

        describe('and user has not yet reached item limit', () => {
          it('should render two info boxes', () => {
            const items = createFakeItems(MAX_WISHLIST_ITEMS - 1)
            const { wrapper } = renderComponent({
              ...props,
              isUserAuthenticated: true,
              items,
              totalItems: items.length,
            })
            const infoBoxes = wrapper.find(WishlistLimitInfo)
            expect(infoBoxes.length).toBe(0)
          })
        })

        describe('and user has reached item limit', () => {
          it('should render an info box at the top and bottom if all items are loaded', () => {
            const items = createFakeItems(MAX_WISHLIST_ITEMS)
            const { wrapper } = renderComponent({
              ...props,
              isUserAuthenticated: true,
              items,
              totalItems: MAX_WISHLIST_ITEMS,
            })
            const infoBoxes = wrapper.find(WishlistLimitInfo)
            expect(infoBoxes.length).toBe(2)
            expect(infoBoxes.first().prop('withMarginTop')).toBe(true)
            expect(infoBoxes.last().prop('withMarginTop')).toBe(false)
          })

          it('should only render an info box at the top if not all items are loaded', () => {
            const items = createFakeItems(MAX_WISHLIST_ITEMS - 20)
            const { wrapper } = renderComponent({
              ...props,
              isUserAuthenticated: true,
              items,
              totalItems: MAX_WISHLIST_ITEMS,
            })
            const infoBox = wrapper.find(WishlistLimitInfo)
            expect(infoBox.length).toBe(1)
            expect(infoBox.prop('withMarginTop')).toBe(true)
          })
        })
      })
    })
  })

  describe('@lifecycle', () => {
    beforeEach(jest.clearAllMocks)

    describe('on componentDidMount', () => {
      const oldWindowNavigatorAgent = window.navigator.userAgent
      beforeAll(() => {
        window.scrollTo = jest.fn()
      })

      afterAll(() => {
        Object.defineProperty(global.navigator, 'userAgent', {
          get: () => oldWindowNavigatorAgent,
          configurable: true,
        })
        window.scrollTo.mockResore()
      })

      describe('fetch WL functionality', () => {
        it('should fetch items if `isWishlistSync` is false', () => {
          const { instance } = renderComponent({
            ...props,
            isWishlistSync: false,
          })
          expect(props.getPaginatedWishlist).not.toHaveBeenCalled()
          instance.componentDidMount()
          expect(props.getPaginatedWishlist).toHaveBeenCalled()
        })

        it('should not fetch items if `isWishlistSync` is true', () => {
          const { instance } = renderComponent({
            ...props,
            isWishlistSync: true,
          })
          instance.componentDidMount()
          expect(props.getPaginatedWishlist).not.toHaveBeenCalled()
        })
      })

      describe('preserve scroll functionality', () => {
        it('should dispatch getPaginatedWishlist when wishlist is not synced', () => {
          const { instance } = renderComponent(props)

          instance.componentDidMount()
          expect(instance.props.getPaginatedWishlist).toHaveBeenCalledTimes(1)
        })
        it('should not dispatch getPaginatedWishlist when wishlist is synced', () => {
          const { instance } = renderComponent({
            ...props,
            isWishlistSync: true,
          })

          instance.componentDidMount()
          expect(instance.props.getPaginatedWishlist).not.toHaveBeenCalled()
        })

        it('should call scrollTo and pass `0,0` as arguments if no IE11', () => {
          Object.defineProperty(global.navigator, 'userAgent', {
            get: () => 'Chrome',
            configurable: true,
          })

          const newProps = { ...props, location: { action: 'SOMETHING' } }
          global.process.browser = true

          const mountComponent = buildComponentRender(
            compose(mountRender),
            WishlistPageContainer
          )
          mountComponent(newProps)
          expect(window.scrollTo).toHaveBeenCalledWith(0, 0)
        })

        it('should call scrollTo and pass `0,preservedScroll` as arguments if IE11', () => {
          Object.defineProperty(global.navigator, 'userAgent', {
            get: () => 'Trident',
            configurable: true,
          })
          const newProps = { ...props, location: { action: 'SOMETHING' } }
          global.process.browser = true

          const mountComponent = buildComponentRender(
            compose(mountRender),
            WishlistPageContainer
          )
          mountComponent(newProps)

          expect(window.scrollTo).toHaveBeenCalledWith(0, 3)
        })
      })
      // @TODO write tests to cover window.toScroll logic
    })

    describe('on UNSAFE_componentWillReceiveProps', () => {
      it('should fetch more items when at the bottom of the current page if not all items loaded', () => {
        const { wrapper } = renderComponent({
          ...props,
          isLoadingItems: false,
          hasReachedPageBottom: false,
          loadedPages: 1,
          totalItems: 100,
          items: createFakeItems(20),
        })

        expect(props.getPaginatedWishlist).not.toHaveBeenCalled()

        wrapper.setProps({ hasReachedPageBottom: true })

        expect(props.getPaginatedWishlist).toHaveBeenCalledWith(2)
      })

      it('should not more items if all items have been loaded', () => {
        const { wrapper } = renderComponent({
          ...props,
          isLoadingItems: false,
          hasReachedPageBottom: false,
          loadedPages: 5,
          totalItems: 100,
          items: createFakeItems(100),
        })

        wrapper.setProps({ hasReachedPageBottom: true })

        expect(props.getPaginatedWishlist).not.toHaveBeenCalled()
      })

      it('should not more items if a request for more items is already in flight', () => {
        const { wrapper } = renderComponent({
          ...props,
          isLoadingItems: true,
          hasReachedPageBottom: false,
          loadedPages: 1,
          totalItems: 100,
          items: createFakeItems(20),
        })

        wrapper.setProps({ hasReachedPageBottom: true })

        expect(props.getPaginatedWishlist).not.toHaveBeenCalled()
      })

      it('should try refetch the first page, if number of wl items changes', () => {
        const { wrapper } = renderComponent(props)

        wrapper.setProps({ totalItems: 52 })

        expect(props.getPaginatedWishlist).toHaveBeenCalledTimes(1)
        expect(props.getPaginatedWishlist).not.toHaveBeenCalledWith(2)
      })
    })
  })

  describe('@events', () => {
    let oldProcessBrowser
    beforeAll(() => {
      oldProcessBrowser = process.browser
      process.browser = true
    })
    afterAll(() => {
      process.browser = oldProcessBrowser
    })

    describe('when the user is not authenticated and the user clicks on the prompt to sign in', () => {
      it('should dispatch triggerWishlistLoginModal', () => {
        const { wrapper } = renderComponent({
          ...props,
          isUserAuthenticated: false,
        })
        expect(props.triggerWishlistLoginModal).not.toHaveBeenCalled()
        wrapper.find(EmptyWishlist).prop('onSignInHandler')()
        expect(props.triggerWishlistLoginModal).toHaveBeenCalledTimes(1)
        expect(props.triggerWishlistLoginModal).toHaveBeenCalledWith(
          null,
          WishlistLoginModal,
          'wishlist'
        )
      })
    })
  })
})
