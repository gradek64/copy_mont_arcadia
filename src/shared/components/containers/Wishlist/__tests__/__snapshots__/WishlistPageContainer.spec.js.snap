// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishlistPageContainer @renders on client when user has items in the wishlist renders <WishlistItemContainer /> CSSTransition 1`] = `
<div
  className="WishlistPageContainer"
>
  <h1
    className="WishlistPageContainer-title"
  >
    My Wishlist
  </h1>
  <span
    className="WishlistPageContainer-itemCount"
  >
    100 items
  </span>
  <CSSTransition
    classNames="fade"
    in={false}
    timeout={
      Object {
        "enter": 400,
        "exit": 0,
      }
    }
    unmountOnExit={true}
  >
    <EmptyWishlist
      isUserAuthenticated={true}
      onSignInHandler={[Function]}
      visited={Array []}
    />
  </CSSTransition>
  <WishlistLimitInfo
    withMarginTop={true}
  />
  <CSSTransition
    classNames="fade"
    in={true}
    timeout={400}
    unmountOnExit={true}
  >
    <Connect(WishlistItemContainer)
      grid={4}
      items={
        Array [
          Object {
            "catEntryId": "31621890",
          },
          Object {
            "catEntryId": "31621847",
          },
        ]
      }
    />
  </CSSTransition>
</div>
`;

exports[`WishlistPageContainer @renders on client when user is authenticated but has no wishlisted items renders <EmptyWishlist /> CSSTransition 1`] = `
<div
  className="WishlistPageContainer"
>
  <h1
    className="WishlistPageContainer-title"
  >
    My Wishlist
  </h1>
  <CSSTransition
    classNames="fade"
    in={true}
    timeout={
      Object {
        "enter": 400,
        "exit": 0,
      }
    }
    unmountOnExit={true}
  >
    <EmptyWishlist
      isUserAuthenticated={true}
      onSignInHandler={[Function]}
      visited={Array []}
    />
  </CSSTransition>
  <CSSTransition
    classNames="fade"
    in={false}
    timeout={400}
    unmountOnExit={true}
  >
    <Connect(WishlistItemContainer)
      grid={4}
      items={Array []}
    />
  </CSSTransition>
</div>
`;

exports[`WishlistPageContainer @renders on client when user is not authenticated renders <EmptyWishlist /> CSSTransition 1`] = `
<div
  className="WishlistPageContainer"
>
  <h1
    className="WishlistPageContainer-title"
  >
    My Wishlist
  </h1>
  <CSSTransition
    classNames="fade"
    in={true}
    timeout={
      Object {
        "enter": 400,
        "exit": 0,
      }
    }
    unmountOnExit={true}
  >
    <EmptyWishlist
      isUserAuthenticated={false}
      onSignInHandler={[Function]}
      visited={Array []}
    />
  </CSSTransition>
  <CSSTransition
    classNames="fade"
    in={false}
    timeout={400}
    unmountOnExit={true}
  >
    <Connect(WishlistItemContainer)
      grid={4}
      items={null}
    />
  </CSSTransition>
</div>
`;
